test/node-api/test_instance_data/test_instance_data.c

```diff
@@ -0,0 +1,236 @@
+    optional_ref = &data->js_cb_ref;
+  }
+
+  NAPI_CALL_RETURN_VOID(env, napi_get_reference_value(env,
+                                                      *optional_ref,
+                                                      &js_cb));
+  NAPI_CALL_RETURN_VOID(env, napi_get_undefined(env, &undefined));
+  NAPI_CALL_RETURN_VOID(env, napi_call_function(env,
+                                                undefined,
+                                                js_cb,
+                                                0,
+                                                NULL,
+                                                NULL));
+  NAPI_CALL_RETURN_VOID(env, napi_delete_reference(env, *optional_ref));
+
+  *optional_ref = NULL;
+}
+
+static void AsyncWorkCbComplete(napi_env env,
+                                   napi_status status,
+                                   void* data) {
+  (void) status;
+  (void) data;
+  call_cb_and_delete_ref(env, NULL);
+}
+
+static bool establish_callback_ref(napi_env env, napi_callback_info info) {
+  AddonData* data;
+  size_t argc = 1;
+  napi_value js_cb;
+
+  NAPI_CALL_BASE(env, napi_get_instance_data(env, (void**)&data), false);
+  NAPI_ASSERT_BASE(env,
+                   data->js_cb_ref == NULL,
+                   "reference must be NULL",
+                   false);
+  NAPI_CALL_BASE(env,
+                 napi_get_cb_info(env, info, &argc, &js_cb, NULL, NULL),
+                 false);
+  NAPI_CALL_BASE(env,
+                 napi_create_reference(env, js_cb, 1, &data->js_cb_ref),
+                 false);
+
+  return true;
+}
+
+static napi_value AsyncWorkCallback(napi_env env, napi_callback_info info) {
+  if (establish_callback_ref(env, info)) {
+    napi_value resource_name;
+    napi_async_work work;
+
+    NAPI_CALL(env, napi_create_string_utf8(env,
+                                           "AsyncIncrement",
+                                           NAPI_AUTO_LENGTH,
+                                           &resource_name));
+    NAPI_CALL(env, napi_create_async_work(env,
+                                          NULL,
+                                          resource_name,
+                                          AsyncWorkCbExecute,
+                                          AsyncWorkCbComplete,
+                                          NULL,
+                                          &work));
+    NAPI_CALL(env, napi_queue_async_work(env, work));
+  }
+
+  return NULL;
+}
+
+static void TestBufferFinalizerCallback(napi_env env, void* data, void* hint) {
+  (void) data;
+  (void) hint;
+  call_cb_and_delete_ref(env, NULL);
+}
+
+static napi_value TestBufferFinalizer(napi_env env, napi_callback_info info) {
+  napi_value buffer = NULL;
+  if (establish_callback_ref(env, info)) {
+    NAPI_CALL(env, napi_create_external_buffer(env,
+                                               sizeof(TestBufferFinalizer),
```

Comment:
I don't think this is actually legal C, `sizeof(SomeFunctionName)`.

Reply:
Well, I *did* run make test *shrug*. I figure it'd grab the size of the function pointer. I can go with `sizeof(napi_callback)`.

Reply:
What I mean is that even if some compilers accept it, not all will (or do the right thing.)

Reply:
OK. I'm changing it to `sizeof(napi_callback)`.